Console input/output procedures.

cheatfate opened the pull request:
Different console procedures which are missing or not properly implemented in Nim' stdlib.
Current PR introduces cross-platform password reading procedure which:
1. Supports reading of UTF-8 passwords. (Does not supported on Nim 1.2.6 stdlib).
2. Supports reading from both console and redirected stdin. (Does not supported on Nim devel stdlib).
3. Supports password size limitation (Removes attack vector which is present in Nim 1.2.6/devel stdlib).
4. Do not raise exceptions (Uses io2.IoResult[T]).

cheatfate committed:
Initial commit.

cheatfate committed:
Fix error handler.

cheatfate committed:
User proper method of console identification (msvcrt).

cheatfate committed:
Use proper method of console identification (glibc).

zah commented on stew/conio.nim:
```diff
@@ -0,0 +1,337 @@
+elif defined(posix):
+  import posix, termios
+
+  proc isConsoleRedirected(consoleFd: cint): bool =
```
In Chronicles, I too need to check if `stdin` is redirected. Let's export something that would allow me perform this check in cross-platform way.

zah commented on stew/conio.nim:
```diff
@@ -0,0 +1,337 @@
+      buffer.setLen(int(min(bytesRead, uint32(maxBytes))))
+
+      # Trim CR/CRLF from buffer.
+      if len(buffer) > 0:
```
This looks like duplicating `strutils.stripLineEnd`.

cheatfate replied:
The problem of `strutils.stripLineEnd` is that it thinks that `\v` and `\f` are line end, but this is not a proper markers of end of line.

zah replied:
These characters are quite obscure nowadays, but OK - if this is a general purpose procedure for reading a line of characters from the screen, why should it preserve a page break character at the end of the line, but not a new line character in the same place?

cheatfate replied:
@zah we are supposed to read UTF-8 encoded strings are you sure `\v` and `\f` could not be part of UTF-8 encoded symbol which was cut because of buffer limits?

zah replied:
Yes, in UTF-8, the first 128 characters in the ASCII table are encoded with a single byte that has a leading zero bit. All other characters are encoded with a sequence of bytes where each byte has a leading bit set to 1. More details here:

https://en.wikipedia.org/wiki/UTF-8#Encoding

cheatfate replied:
@zah  you are right, but i still think `\v` and `\f` is not a proper condition for line end which we are going to receive from console. When we reading from console the only proper line end is `\n` and `\r\n` and nothing else, `strutils` procedure is incorrect.

zah commented on stew/conio.nim:
```diff
@@ -0,0 +1,337 @@
+      # Truncate additional bytes from buffer.
+      buffer.setLen(int(min(charsRead, uint32(maxBytes))))
+      # Truncate CRLF in result wide string.
+      if len(buffer) > 0:
```
Ditto (`strutils.stripLineEnd`)

cheatfate replied:
This works on `int16` chunks, not on chars/bytes.

cheatfate committed:
Make isConsoleRedirected() public API.

cheatfate committed:
Add UTF-8 validation procedure.

cheatfate committed:
Add UTF-8 length procedure.
Add UTF-8 offset procedure.
Add UTF-8 substr procedure.
Add wchar_t to UTF-8 conversion procedure.
Add multibyte to wchar_t conversion procedure (posix).
Add UTF-8 tests.
Fix password reader to validate utf-8 encoding when reading from pipe.
Fix password reader to read utf-8 encoded strings from *nix console.

cheatfate committed:
Fix UTF-32 encoder/decoder.
Add tests for UTF-8 to UTF-32 and UTF-32 to UTF-8 encoders.

cheatfate committed:
Fix *nix compilation problem.

jangko commented:
I think we can extract the utf8 module + tests and combine it with my utf-8/16 module. we have some overlap functionality and some are distinct.

arnetheduck commented on stew/utf8.nim:
```diff
@@ -0,0 +1,466 @@
+
+type
+  UResult*[T] = Result[T, cstring]
+  Wides32* = int32 | uint32
```
`std/unicode` and `Rune`? this way, it would integrate with other nim libraries

arnetheduck commented on stew/utf8.nim:
```diff
@@ -0,0 +1,466 @@
+type
+  UResult*[T] = Result[T, cstring]
+  Wides32* = int32 | uint32
+  Wides16* = int16 | uint16
```
unused

arnetheduck commented on stew/utf8.nim:
```diff
@@ -0,0 +1,466 @@
+  ErrorInvalidLocale* = cstring"Could not obtain system locale"
+  ErrorNotEnoughCharacters* = cstring"Not enough characters in string"
+
+proc utf8Validate*[T: Bytes](data: openarray[T]): bool =
```
std/unicde.validateUTF8?

tersec commented:
Why do the UTF-8/32 parts?
